crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move

```diff
@@ -18,4 +21,222 @@ module sui::bls12381 {
+    const SCALAR_TYPE: u8 = 2;
+    const G1_TYPE: u8 = 3;
+    const G2_TYPE: u8 = 4;
+    const GT_TYPE: u8 = 5;
+
+    ///////////////////////////////
+    ////// Scalar operations //////
+
+    public fun scalar_from_bytes(bytes: &vector<u8>): Element<Scalar> {
+        group_ops::from_bytes(SCALAR_TYPE, bytes, false)
+    }
+
+    public fun scalar_from_u64(x: u64): Element<Scalar> {
+        let bytes = SCALAR_ZERO_BYTES;
+        group_ops::set_as_prefix(x, true, &mut bytes);
+        group_ops::from_bytes(SCALAR_TYPE, &bytes, true)
+    }
+
+    public fun scalar_zero(): Element<Scalar> {
+        group_ops::from_bytes(SCALAR_TYPE, &SCALAR_ZERO_BYTES, true)
+    }
+
+    public fun scalar_one(): Element<Scalar> {
+        group_ops::from_bytes(SCALAR_TYPE, &SCALAR_ONE_BYTES, true)
+    }
+
+    public fun scalar_add(e1: &Element<Scalar>, e2: &Element<Scalar>): Element<Scalar> {
+        group_ops::add(SCALAR_TYPE, e1, e2)
+    }
+
+    public fun scalar_sub(e1: &Element<Scalar>, e2: &Element<Scalar>): Element<Scalar> {
+        group_ops::sub(SCALAR_TYPE, e1, e2)
+    }
+
+    public fun scalar_mul(e1: &Element<Scalar>, e2: &Element<Scalar>): Element<Scalar> {
+        group_ops::mul(SCALAR_TYPE, e1, e2)
+    }
+
+    // Returns e2/e1, fails if a is zero.
+    public fun scalar_div(e1: &Element<Scalar>, e2: &Element<Scalar>): Element<Scalar> {
+        group_ops::div(SCALAR_TYPE, e1, e2)
+    }
+
+    public fun scalar_neg(e: &Element<Scalar>): Element<Scalar> {
+        scalar_sub(&scalar_zero(), e)
+    }
+
+    // Fails if e is zero.
+    public fun scalar_inv(e: &Element<Scalar>): Element<Scalar> {
+        scalar_div(e, &scalar_one())
+    }
+
+    // TODO: msm? is it useful?
+
+    /////////////////////////////////
+    ////// G1 group operations //////
+
+    public fun g1_from_bytes(bytes: &vector<u8>): Element<G1> {
+        group_ops::from_bytes(G1_TYPE, bytes, false)
+    }
+
+    public fun g1_identity(): Element<G1> {
+        group_ops::from_bytes(G1_TYPE, &G1_IDENTITY_BYTES, true)
+    }
+
+    public fun g1_generator(): Element<G1> {
+        group_ops::from_bytes(G1_TYPE, &G1_GENERATOR_BYTES, true)
+    }
+
+    public fun g1_add(e1: &Element<G1>, e2: &Element<G1>): Element<G1> {
+        group_ops::add(G1_TYPE, e1, e2)
+    }
+
+    public fun g1_sub(e1: &Element<G1>, e2: &Element<G1>): Element<G1> {
+        group_ops::sub(G1_TYPE, e1, e2)
+    }
+
+    public fun g1_mul(e1: &Element<Scalar>, e2: &Element<G1>): Element<G1> {
+        group_ops::mul(G1_TYPE, e1, e2)
+    }
+
+    // Returns e2 / e1, fails if scalar is zero.
+    public fun g1_div(e1: &Element<Scalar>, e2: &Element<G1>): Element<G1> {
+        group_ops::div(G1_TYPE, e1, e2)
+    }
+
+    public fun g1_neg(e: &Element<G1>): Element<G1> {
+        g1_sub(&g1_identity(), e)
+    }
+
+    // Hash using DST = BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_
+    public fun hash_to_g1(m: &vector<u8>): Element<G1> {
+        group_ops::hash_to(G1_TYPE, m)
+    }
```

Comment:
One way to potentially clean this up is to turn these into native functions that are generic on the element type, e.g.:

```move
    public native fun from_bytes<G>(bytes: &vector<u8>): Element<G>;
    public native fun identity<G>(): Element<G>;
    public native fun generator<G>(): Element<G>;
    public native fun add<G>(e1: &Element<G>, e2: &Element<G>): Element<G>;
    public native fun sub<G>(e1: &Element<G>, e2: &Element<G>): Element<G>;
    public native fun mul<G>(e1: &Element<Scalar>, e2: &Element<G>): Element<G>;
    public native fun div<G>(e1: &Element<Scalar>, e2: &Element<G>): Element<G>;

    public fun neg<G>(e: &Element<G>): Element<G> {
        sub(&identity(), e)
    }

    public native fun hash_to<G>(m: &vector<u8>): Element<G>;
```

Then in the native function implementation, you dispatch the implementation based on the type parameter (particularly important for `identity` and `generator` whose bytes will presumably differ based on which group type you are using.

Reply:
Thanks for this proposal. The disadvantage with this option was that some user mistakes are only detectable in runtime (e.g., if I call `div` with Scalar type that does not work with G, or if I just call any of the functions with an unsupported type).